x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -314,48 +316,45 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
-                            if (resolvedIndices.isNoIndicesPlaceholder()) {
-                                listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.ALLOW_NO_INDICES));
-                            } else {
-                                buildIndicesAccessControl(authentication, action, authorizationInfo,
-                                    Sets.newHashSet(resolvedIndices.getLocal()), aliasOrIndexLookup, listener);
-                            }
-                        }, listener::onFailure));
-                    } else {
-                        listener.onResponse(indexAuthorizationResult);
-                    }
-                }, listener::onFailure));
+                        //all wildcard expressions have been resolved and only the security plugin could have set '-*' here.
+                        //'-*' matches no indices so we allow the request to go through, which will yield an empty response
+                        if (resolvedIndices.isNoIndicesPlaceholder()) {
+                            listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.ALLOW_NO_INDICES));
+                        } else {
+                            listener.onResponse(buildIndicesAccessControl(
+                                action, authorizationInfo, Sets.newHashSet(resolvedIndices.getLocal()), aliasOrIndexLookup));
+                        }
+                    }, listener::onFailure));
+                } else {
+                    listener.onResponse(indexAuthorizationResult);
+                }
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
         }
     }
 
-    private void authorizeIndexActionName(String action, AuthorizationInfo authorizationInfo, IndicesAccessControl grantedValue,
-                                          ActionListener<IndexAuthorizationResult> listener) {
-        if (authorizationInfo instanceof RBACAuthorizationInfo) {
-            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
-            if (role.checkIndicesAction(action)) {
-                listener.onResponse(new IndexAuthorizationResult(true, grantedValue));
-            } else {
-                listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.DENIED));
-            }
-        } else {
-            listener.onFailure(new IllegalArgumentException("unsupported authorization info:" +
-                authorizationInfo.getClass().getSimpleName()));
-        }
+    private static IndexAuthorizationResult authorizeIndexActionName(String action,
+                                                                     AuthorizationInfo authorizationInfo,
+                                                                     IndicesAccessControl grantedValue) {
+        ensureRBAC(authorizationInfo);
+        final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
```

Comment:
Nit: I think `ensureRBAC` can do the cast and return a `RBACAuthorizationInfo` object so that caller no longer needs to do it.

Reply:
++